### PR TITLE
Separated crossjoin from hashjoin & Optimized it in-memory

### PR DESCRIPTION
**Pull request by watemus**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Separated CrossJoin from HashJoin

- Optimized it in-memory (will move data to disk if the memory limit is exceeded)

### Documentation entry for user-facing changes

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit d70dd5b8b9af61812e65d06d20b3f6ed48c7ceb1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50057/d70dd5b8b9af61812e65d06d20b3f6ed48c7ceb1/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🔴 failure</td></tr>
</table>

**Review by vdimir — COMMENTED** [review]
Thank you for your pull request!
Overall, it looks good.

The main concern is regarding how the single stream handles the multi-threaded joinBlock (see comments). Additionally, it might be beneficial to minimize the amount of reading for the right table if we preserve the reading stream's position.

Do the existing tests for CROSS JOIN cover the new feature with the move to disc scenario? Could you also provide some test cases that you used to measure the performance of the implementation, as well as any results you have obtained, if you did it?

File: src/Interpreters/CrossJoin.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            {
+                const IColumn & column_right = *block_right.getByPosition(col_num).column;
+                dst_columns[num_existing_columns + col_num]->insertRangeFrom(column_right, 0, rows_right);
+            }
+        };
+
+        Block block_right;
+        do
+        {
+            block_right = blockStream.read(false);
```
[the remark this answers is not in the document]
Comment (watemus) [reply src/Interpreters/CrossJoin.cpp]: fixed

File: src/Interpreters/CrossJoin.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+{
+    if (unlikely(block.rows() > std::numeric_limits<RowRef::SizeT>::max()))
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Too many rows in right table block for HashJoin: {}", block.rows());
+
+    Block materialized = materializeBlock(block);
+
+    ++right_blocks_count;
+    right_blocks.push_back(std::move(materialized));
+    right_rows += block.rows();
+    right_bytes += block.bytes();
```
[the remark this answers is not in the document]
Comment (watemus) [reply src/Interpreters/CrossJoin.cpp]: fixed

**Comment by vdimir** [comment]
I'll close it for now and reopen or cherry-pick it when we have a chance to continue.

**Pull request closed without merging**